### ISSUE TITLE
Enum doesn't compile - "method values needs result type because its right-hand side attempts implicit search"

### ISSUE THREAD

**Issue by ysthakur** [lampepfl/dotty#8797]
I was trying out Dotty's enums in a project of mine, and everything was going fine, until one day I started getting the error `"method values needs result type because its right-hand side attempts implicit search ... To avoid this error, give 'method values' an explicit type."`

At first, I could get rid of this by running sbt clean before compile, but I recently moved my enum to a new module and now running clean beforehand doesn't make the error go away.

I have no idea why this is happening, since I haven't changed my Dotty version (0.22.0) and just moving the file somewhere should have no effect on anything. 

At first I thought it was because the values method didn't have a return type declared when being generated, so I typed this below my enum `SymbolTokenType`:
```
object SymbolTokenType {
  def values: Array[SymbolTokenType] = ???
}
```
and I got an error saying that a method with the same signature was already defined, which means that dotc *does* give a proper return type to method values. Or am I interpreting "result type" wrong?

**Comment by odersky** [comment]
An enum will generate a `values` method. My guess is that this interacted in some way with a user-defined values method in your code. The error you saw comes from a cyclic reference that was detected by the compiler. Hard to say more without knowing the code.

#8814 changes the code generation so that `values` has an explicit result type. That should either fix the problem in your code or give you a different error message.

**Comment by ysthakur** [comment]
Apparently, I had an implicit def in scope that operated on instances of TokenType, which was extended by SymbolTokenType. That may have been what caused the problem.

**Comment by ysthakur** [comment]
This wouldn't work with 0.24.0-RC1 either, but it did start working with a small tweak (although I'm not sure why exactly). Turns out this isn't a bug with Dotty, just something dumb I did.

I had `trait TokenType` and its children SymbolTokenType, KeywordTokenType, etc. and also `object TokenType` in TokenType.scala. Inside the object `TokenType`, I had the following export statements so I could import all the different `TokenType` cases at once:

```
export com.ysthakur.blah.SymbolTokenType._
export com.ysthakur.blah.KeywordTokenType._
```

Removing the export statements solved it, though I'm not sure exactly why, because I didn't have anything like `import TokenType._` inside `TokenType.scala`. All the values of SymbolTokenType that were exported from the companion object shouldn't have been visible inside `SymbolTokenType` or inside the package `com.ysthakur.blah`.

Once I moved the companion object `TokenType` into a separate file (in the same package), it magically started working again, though.
